### PR TITLE
Solve day3

### PR DESCRIPTION
**Pull request by haynaou**

### CODE REVIEW

**Review by zianwar — APPROVED** [review]

File: day3b.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+maybe_part_numbers = find_maybe_part_numbers(lines)
+total = 0
+neighbor_part_number_map = defaultdict(list)
+for maybe_part_number in maybe_part_numbers:
+    neighbors = get_neighbor_positions(lines, maybe_part_number)
+    for neighbor in neighbors:
+        neighbor_part_number_map[neighbor].append(maybe_part_number)
+    if has_symbol(lines, neighbors):
+        total += maybe_part_number.value
```
Comment (shadanan) [inline day3b.py]: This as well.
```suggestion
```

File: day3b.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    return valid_neighbors
+
+def has_symbol(lines, neighbors: list[tuple[int, int]]) -> bool:
+    for r, c in neighbors:
+        if not lines[r][c].isnumeric() and lines[r][c] != ".":
+            return True
+    return False
+
+maybe_part_numbers = find_maybe_part_numbers(lines)
+total = 0
```
Comment (shadanan) [inline day3b.py]: This is no longer needed.
```suggestion
```

File: day3b.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    end: int
+    value: int
+
+
+def find_maybe_part_numbers(lines: list[str]) -> list[PartNumber]:
+    nums = []
+    for row, line in enumerate(lines):
+        for match in NUMBER_PATERN.finditer(line):
+            nums.append(PartNumber(row, match.start(), match.end(), int(match.group())))
+    return nums
```
Comment (shadanan) [inline day3b.py]: We should name `nums` -> `part_numbers`:
```suggestion
    part_numbers = []
    for row, line in enumerate(lines):
        for match in NUMBER_PATERN.finditer(line):
            part_numbers.append(PartNumber(row, match.start(), match.end(), int(match.group())))
    return part_numbers
```

File: day3b.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+@dataclass
+class PartNumber:
+    row: int
+    start: int
+    end: int
+    value: int
+
+
+def find_maybe_part_numbers(lines: list[str]) -> list[PartNumber]:
```
Comment (shadanan) [inline day3b.py:19]: A thought about how to make this code more readable:

1. Rename `find_maybe_part_numbers` -> `find_part_numbers`.
2. Do what @zianwar was suggesting -- combine `get_neighbor_positions` and `has_symbol` into a single function called `is_valid`.
3. Call `is_valid` at line 23, passing the part number.

**Review by shadanan — APPROVED** [review]

**Comment by shadanan** [comment]
@haynaou - I noticed that in https://github.com/shadanan/advent-of-code-2023/pull/1 you applied my suggestions. FYI, in general, suggestions that reviewers make will typically be untested. You'll want to apply the changes, pull them locally, and re-run the code to make sure that it works before merging.

**Pull request merged**